storage/src/main/java/io/delta/storage/internal/ThreadUtils.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.storage.internal;
+
+import java.io.InterruptedIOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public final class ThreadUtils {
+
+    /**
+     * Based on Apache Spark's ThreadUtils.runInNewThread
+     * Run a piece of code in a new thread and return the result.
+     */
+    public static <T> T runInNewThread(
+            String threadName,
+            boolean isDaemon,
+            Callable<T> body) throws Throwable {
+        // Using a single-element list to hold the throwable and result,
+        // since values used in static method must be final
+        List<Throwable> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
+        Thread thread = new Thread(threadName) {
+            @Override
+            public void run() {
+                try {
+                    resultHolder.add(body.call());
+                } catch (Throwable t) {
+                    if (LogStoreErrors.isNonFatal(t)) {
+                        exceptionHolder.add(t);
+                    }
+                }
+            }
+        };
+        thread.setDaemon(isDaemon);
+        thread.start();
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+            throw new InterruptedIOException(e.getMessage());
+        }
```

Comment:
Move this to the caller. `runInNewThread` may run non IO code then `InterruptedIOException` doesn't make sense in this case.

Reply:
decided to just remove, and not move to the caller. we are catching and casting as a runtime exception instead.

Reply:
> we are catching and casting as a runtime exception instead.

It's better to avoid casting `InterruptedException` to a runtime exception. For example, Streaming queries will check `InterruptedException` or `InterruptedIOException`, and use them as a normal signal to stop a streaming query. Changing it to a runtime exception will make the streaming query go into the failure state.

Reply:
so re-throw it as an InterruptedIOException?